### PR TITLE
Executor work

### LINKED ISSUE #48: Executor development and design
The new executor design document is here:
https://paper.dropbox.com/doc/Bastion-Executor-Design-Document-AmsdBId9x5CDqL5ewJT0IroAg-ged3kQPsVFNOIwuhgkeIY

This is a bigger feature that needs heavy collab from both runtime and also proc side.

## Main requirements
- [x] Fair scheduling with multiqueue
- [x] Optimal performance distribution with minimal cache misses
- [ ] Thread recovery

### PR DESCRIPTION
**Pull request by vertexclique**
Bastion's Executor Development

https://paper.dropbox.com/doc/Bastion-Executor-Design-Document--AnZMiua4Egf6vE_pXCH_2RGQAg-ged3kQPsVFNOIwuhgkeIY

Fixes #48

### CODE REVIEW

File: bastion-executor/src/lib.rs [hunk 1]
```diff
@@ -1,2 +1,35 @@
...
+
+// Executing asm in some places
+#![feature(asm)]
+// Allocator features
+#![feature(allocator_api)]
+#![feature(extern_types)]
+#![feature(core_intrinsics)]
+#![feature(libstd_sys_internals)]
+#![feature(thread_local)]
+#![feature(const_fn)]
```
Comment (azastrael) [inline bastion-executor/src/lib.rs]: Can be removed ;)

File: bastion-executor/src/placement.rs [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+#[cfg(target_os = "windows")]
+#[inline]
+fn set_for_current_helper(core_id: CoreId) {
+    windows::set_for_current(core_id);
+}
+
+#[cfg(target_os = "windows")]
+extern crate kernel32;
+#[cfg(target_os = "windows")]
+extern crate winapi;
```
Comment (azastrael) [inline bastion-executor/src/placement.rs]: Can probably be removed

File: bastion-executor/src/pool.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+pub struct Pool {
+    pub injector: Injector<LightProc>,
+    pub stealers: Vec<Stealer<LightProc>>,
+    pub sleepers: Sleepers,
+}
+
+impl Pool {
+    /// Error recovery for the fallen threads
+    pub fn recover_async_thread() {
+        unimplemented!()
```
Comment (azastrael) [inline bastion-executor/src/pool.rs]: Shoud we create an issue or put a `// FIXME`? :)

File: bastion-executor/src/run.rs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        // Log this `block_on` operation.
+        let _child_id = stack.get_pid();
+        let _parent_id = worker::get_proc_stack(|t| t.get_pid()).unwrap_or(0);
+
+        // Wrap the future into one that drops task-local variables on exit.
+        // let future = task_local::add_finalizer(future);
+
+        let future = async move {
+            future.await;
+        };
```
Comment (azastrael) [inline bastion-executor/src/run.rs]: Can be removed ;)

File: bastion-executor/src/run.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                *out = Some(future.await);
+            }
+        };
+
+        // Log this `block_on` operation.
+        let _child_id = stack.get_pid();
+        let _parent_id = worker::get_proc_stack(|t| t.get_pid()).unwrap_or(0);
+
+        // Wrap the future into one that drops task-local variables on exit.
+        // let future = task_local::add_finalizer(future);
```
Comment (azastrael) [inline bastion-executor/src/run.rs]: Can this be removed? :)

File: bastion-executor/src/thread_recovery.rs [hunk 6]
```diff
@@ -0,0 +1 @@
+
```
Comment (azastrael) [inline bastion-executor/src/thread_recovery.rs]: This file's empty ;)

File: bastion-executor/src/distributor.rs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            last_dead: usize::max_value(),
+            cores: placement::get_core_ids().expect("Core mapping couldn't be fetched"),
+        }
+    }
+
+    pub fn assign(mut self) -> Vec<Stealer<LightProc>> {
+        let mut stealers = Vec::<Stealer<LightProc>>::new();
+
+        for core in self.cores {
+            self.round = core.id;
```
Comment (azastrael) [inline bastion-executor/src/distributor.rs]: `Distributor.round` doesn't seem to be used anywhere else

File: bastion-executor/src/blocking_pool.rs [hunk 8]
```diff
@@ -0,0 +1 @@
+
```
Comment (azastrael) [inline bastion-executor/src/blocking_pool.rs]: This file's empty ;)

**Comment by vertexclique** [comment]
We are approximately ready to glue bastion + executor + lightproc.
This is going in for now!

**Pull request merged**